In [7]:
from sklearn.model_selection import train_test_split
from sklearn.datasets import load_diabetes
import pandas as pd

data = load_diabetes()
X = pd.DataFrame(data.data, columns=data.feature_names)
y = data.target

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

print('Размер тренировочных данных:', X_train.shape)
print('Размер тестовых данных:', X_test.shape)

Размер тренировочных данных: (353, 10)
Размер тестовых данных: (89, 10)


In [8]:
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error
import numpy as np

model = LinearRegression()
model.fit(X_train, y_train)

predictions = model.predict(X_test)
mse = mean_squared_error(y_test, predictions)
print('RMSE:', np.sqrt(mse))

RMSE: 53.85344583676593


In [9]:
print(y.min(), y.max())

25.0 346.0


Модель ошибается в среднем на 53.85 единицы. Учитывая что диапазон целевой переменной составляет 25-346 (~320 единиц), ошибка составляет около 17% — приемлемый результат для линейной регрессии на медицинских данных.

In [12]:
from sklearn.model_selection import cross_val_score
scores = cross_val_score(model, X, y, cv=5)
print(scores.mean())

0.48231643590864215


Кросс-валидация с 5 фолдами дала R² = 0.48. Это означает что модель объясняет лишь 48% разброса в данных. Результат слабый — модель предсказывает прогрессирование диабета лишь немного лучше чем простое среднее. Для улучшения можно попробовать другие алгоритмы или добавить больше признаков.